feat: Support for vectors creation when adding  records from Python SDK

frascuchon opened the pull request:
<!-- Thanks for your contribution! As part of our Community Growers initiative 🌱, we're donating Justdiggit bunds in your name to reforest sub-Saharan Africa. To claim your Community Growers certificate, please contact David Berenstein in our Slack community or fill in this form https://tally.so/r/n9XrxK once your PR has been merged. -->

# Description

This PR allows Python SDK users to provide vectors when adding records to a feedback dataset. Also, the default error handler mechanism has been fixed to include more info on API errors

This [PR](https://github.com/argilla-io/argilla/pull/4055) must be merged first

Refs [#3028](https://github.com/argilla-io/argilla/issues/3028)

**Type of change**

(Please delete options that are not relevant. Remember to title the PR according to the type of change)

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactor (change restructuring the codebase without changing functionality)
- [X] Improvement (change adding some improvement to an existing functionality)

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes. And ideally, reference `tests`)

- [ ] Test A
- [ ] Test B

**Checklist**

- [ ] I added relevant documentation
- [ ] I followed the style guidelines of this project
- [ ] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [ ] I have added relevant notes to the `CHANGELOG.md` file (See https://keepachangelog.com/)

frascuchon committed:
feat: Add vectors dict to the FeedbackRecord Model

jfcalvo committed:
fix: failing test

frascuchon committed:
Merge branch 'feat/add-vector-settings-from-python-sdk' into feat/create-records-with-vectors-in-python-sdk

frascuchon committed:
Merge branch 'feat/add-vector-settings-from-python-sdk' into feat/create-records-with-vectors-in-python-sdk

frascuchon committed:
chore: Avoid sending suggestions in requests when not defined at record level

frascuchon committed:
tests: Adding tests for records creation with vectors

frascuchon committed:
chore: Update changelog

frascuchon committed:
tests: Adding test creating records with wrong vectors.

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

frascuchon committed:
Merge branch 'feat/add-vector-settings-from-python-sdk' into feat/create-records-with-vectors-in-python-sdk

frascuchon committed:
revert: Skip suggestions if not provided (empty list is a valid value)

gabrielmbmb commented on src/argilla/client/feedback/schemas/records.py:
we need to add the docstring for the `vectors` attribute of the `FeedbackRecord` class

jfcalvo commented on CHANGELOG.md:
```diff
@@ -36,7 +36,7 @@ These are the section headers that we use:
 - Update `FeedbackDataset.__init__` to allow passing a vector_settings list. ([#4055](https://github.com/argilla-io/argilla/pull/4055))
 - Update `FeedbackDataset.push_to_argilla` to also push vector settings. ([#4055](https://github.com/argilla-io/argilla/pull/4055))
-
+- Update `FeedbackDatasetRecord` to allow to create records with vectors. ([#4043](https://github.com/argilla-io/argilla/pull/4043))
```
```suggestion
- Update `FeedbackDatasetRecord` to support the creation of records with vectors. ([#4043](https://github.com/argilla-io/argilla/pull/4043))
```

jfcalvo commented on tests/integration/client/feedback/dataset/remote/test_dataset.py:
```diff
@@ -452,6 +452,60 @@ def test_adding_vector_setting_with_the_same_name(self, owner: "User", feedback_
+        remote_dataset.add_records(records)
+
+        assert len(remote_dataset.records) == len(records)
+        # TODO: test this once the list of vectors is returned
+        # assert all(record.vectors["vector"] for record in remote_dataset.records)
```
We can do this right? The feature is already merged in the feature branch, the one supporting `include=vectors`. Not sure if that will require a change in the Api function doing the request.

frascuchon replied:
We will modify this test once the include vectors is supported through the Python SDK

jfcalvo commented on tests/integration/client/feedback/dataset/remote/test_dataset.py:
```diff
@@ -452,6 +452,60 @@ def test_adding_vector_setting_with_the_same_name(self, owner: "User", feedback_
+                    fields={"text": "Hello world!", "text-2": "Hello world!"},
+                    vectors=wrong_vectors,
+                )
+            )
```
I think we should use `invalid` instead of `wrong`:

```suggestion
    def test_adding_records_with_invalid_vectors(
        self, owner: "User", feedback_dataset: FeedbackDataset, invalid_vectors: dict, expected_error: str
    ):
        api.init(api_key=owner.api_key)
        workspace = Workspace.create(name="test-workspace")

        vector_dimension = 4
        feedback_dataset.add_vector_settings(VectorSettings(name="vector", dimensions=vector_dimension))
        remote_dataset = feedback_dataset.push_to_argilla(name="test_dataset", workspace=workspace)

        with pytest.raises(ValidationApiError, match=expected_error):
            remote_dataset.add_records(
                FeedbackRecord(
                    fields={"text": "Hello world!", "text-2": "Hello world!"},
                    vectors=invalid_vectors,
                )
            )
```

frascuchon committed:
Update CHANGELOG.md

Co-authored-by: José Francisco Calvo <jose@argilla.io>

frascuchon commented on tests/integration/client/feedback/dataset/remote/test_dataset.py:
```diff
@@ -452,6 +452,60 @@ def test_adding_vector_setting_with_the_same_name(self, owner: "User", feedback_
+        # assert all(record.vectors["vector"] for record in remote_dataset.records)
+
+    @pytest.mark.parametrize(
+        "wrong_vectors, expected_error",
```
``` suggestion
        "invalid_vectors, expected_error",
```

frascuchon committed:
Apply suggestions from code review

Co-authored-by: José Francisco Calvo <jose@argilla.io>

frascuchon committed:
feature: search records using vectors (#4019)

<!-- Thanks for your contribution! As part of our Community Growers
initiative 🌱, we're donating Justdiggit bunds in your name to reforest
sub-Saharan Africa. To claim your Community Growers certificate, please
contact David Berenstein in our Slack community or fill in this form
https://tally.so/r/n9XrxK once your PR has been merged. -->

# Description

This PR updates the endpoint `POST
/api/me/datasets/:dataset_id/records/search` to accept vector searches:
```bash
POST /api/me/datasets/:dataset_id/records/search
{
  "query": {
    "vector": {
      "name": "my-vector-name",
      "record_id": ...,
      "order": "most_similar"
    }
  }
}
````

Also, the minimal OpenSearch version has been upgraded to 2.5, since
versions <=2.4 do not support filtering + similarity search properly.

Finally, the default Argilla search engine is configured to
Elasticsearch (⚠️ this change may affect users using OpenSearch)

## TODO
- [x] Add missing tests
- [x] Filter using metadata + response status 
- [x] Apply most/least similarity order
- [x] Validations (record exists by id)
- [x] Adding search text as filter

PR https://github.com/argilla-io/argilla/pull/4043 must be merged before

Refs: https://github.com/argilla-io/argilla/issues/3028

**Type of change**

(Please delete options that are not relevant. Remember to title the PR
according to the type of change)

- [ ] New feature (non-breaking change which adds functionality)
- [ ] Refactor (change restructuring the codebase without changing
functionality)
- [x] Improvement (change adding some improvement to an existing
functionality)

**How Has This Been Tested**

(Please describe the tests that you ran to verify your changes. And
ideally, reference `tests`)

- [ ] Test A
- [ ] Test B

**Checklist**

- [ ] I added relevant documentation
- [x] I followed the style guidelines of this project
- [ ] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my
feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK)
(see text above)
- [x] I have added relevant notes to the `CHANGELOG.md` file (See
https://keepachangelog.com/)

---------

Co-authored-by: pre-commit-ci[bot] <66853113+pre-commit-ci[bot]@users.noreply.github.com>
Co-authored-by: José Francisco Calvo <jose@argilla.io>

frascuchon committed:
Merge branch 'feat/add-vector-settings-from-python-sdk' into feat/create-records-with-vectors-in-python-sdk

frascuchon committed:
chore: Remove non-required async signature
